File: closest_pair_heuristic/brute_force_closest_pair_heuristic.c
#include <limits.h>
#include <signal.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wmissing-prototypes"
void setUp(void) {}

void tearDown(void) {}
#pragma clang diagnostic pop

/* ClosestPair(P)
 *      Let n be the number of points in set P.
 *      For i = 1 to n − 1 do
 *          d=∞
 *          For each pair of endpoints (s, t) from distinct vertex chains
 *              if dist(s, t) ≤ d then sm = s, tm = t, and d = dist(s, t)
 *          Connect (sm, tm) by an edge
 *      Connect the two endpoints by an edge
 *
 * [source: algo design manual, p.7]
 */

static int dist(int n1, int n2) { /* {{{ */
    return abs(n1 - n2);
}
/* }}} */
static void distinct_vertex_chains(int ret[][2], int ret_len, int const endpoints[], int endpoints_len) { /* {{{ */
    int offset = 0;
    for (int i = 0; i < endpoints_len && offset < ret_len; i++) {
        for (int j = i; j < endpoints_len - 1; j++) {
            ret[offset][0] = endpoints[i];
            ret[offset][1] = endpoints[j + 1];
            offset++;
        }
    }
}
/* }}} */
static void print_pairs(int pairs[][2], int const len) { /* {{{ */
    for (int i = 0; i < len; i++) {
        printf("(%d, %d)\n", pairs[i][0], pairs[i][1]);
    }
}
static int triangular_number(int height) {
    return (height * (height + 1)) / 2;
}
/* }}} */
int main(void) {
    int P[] = {0, -21, 11, -5, 3, -1, 1};
    int n = sizeof(P) / sizeof(int);

    int longest_row_len = n - 1;
    int num_uniq_pairs=triangular_number(/*height=*/longest_row_len);
    int pairs[num_uniq_pairs][2];

    distinct_vertex_chains(/*ret=*/pairs, /*ret_len=*/triangular_number(/*height=*/longest_row_len), /*endpoints=*/P, /*endpoints_len=*/n);
    print_pairs(pairs, /*len=*/num_uniq_pairs);

    for (unsigned long i = 1; i < n; i++) {
         int d = INT_MAX;
        int s;
        int t;
        int sm;
        int tm;
         for (int j=0; j < num_uniq_pairs; j++) {
             s = pairs[j][0];
             t = pairs[j][1];
            if (dist(s, t) <= d) {
                d = dist(s,t);
                sm = s;
                tm = t;
            }
         }
         int edge[2] = {sm, tm};
         printf("(%d, %d)\n", edge[0], edge[1]);
    }
    //         int s_ = pairs[i][0];
    //         int t_ = pairs[j][1];
    //         if (dist(s_, t_) <= d) {
    //             s = s_;
    //             t = t_;
    //             d = dist(s_, t_);
    //         }
    //         printf("j=%lu; (%d, %d); d=%d\n", j, s, t, d);
    //     }
    // }

    return 0;
}

File: closest_pair_heuristic/brute_force_closest_pair_heuristic.py
# pyright: strict

import logging
import typing

# ClosestPair(P)
#      Let n be the number of points in set P.
#      For i = 1 to n − 1 do
#          d=∞
#          For each pair of endpoints (s, t) from distinct vertex chains
#              if dist(s, t) ≤ d then sm = s, tm = t, and d = dist(s, t)
#          Connect (sm, tm) by an edge
#      Connect the two endpoints by an edge
#
# [source: algo design manual, p.7]


def configure_logging() -> None:  # {{{
    logging.basicConfig(
        format="%(asctime)s,%(msecs)03d %(levelname)-8s [%(filename)s:%(lineno)d] %(message)s",
        datefmt="%Y-%m-%d:%H:%M:%S",
        level=logging.DEBUG,
        handlers=[logging.StreamHandler()],
    )


# }}}


def uniq_pairs(P: typing.Sequence[int]) -> list[tuple[int, int]]:  # {{{
    n = len(P)
    ret: list[tuple[int, int]] = []
    for i in range(1, n):
        for j in range(i, n):
            ret.append((P[i - 1], P[j]))
    return ret


# }}}


def dist(s: int, t: int) -> int:  # {{{
    return abs(s - t)


# }}}


def closest_pairs(P: typing.Sequence[int], n: int):  # {{{
    ret: list[tuple[int, int]] = []
    connected_nodes: set[int] = set()

    for _ in range(1, n):

        d = float("inf")
        sm = 0
        tm = 0

        for pair in uniq_pairs(P):

            if pair[0] in connected_nodes and pair[1] in connected_nodes:
                continue

            s = pair[0]
            t = pair[1]

            if dist(s, t) <= d:
                sm = s
                tm = t
                d = dist(s, t)

        connected_nodes.update((sm, tm))
        ret.append((min(sm, tm), max(sm, tm)))

    return ret


# }}}


def print_cycle(edges: typing.Sequence[tuple[int, int]]):  # {{{
    cycle: list[int] = []

    for pair in sorted(edges):

        if not cycle:
            cycle.extend(pair)
            continue

        assert cycle[-1] == pair[0]

        cycle.append(pair[1])

    cycle.append(cycle[0])

    print(cycle)


# }}}


def main():
    configure_logging()
    P: list[int] = [0, -21, 11, -5, 3, -1, 1]
    print_cycle(edges=closest_pairs(P, n=len(P)))


if __name__ == "__main__":
    main()
